### DISCUSSION TITLE
Possible Event Pool refactor

### DISCUSSION THREAD

**Discussion by danielalejandromatheus** [m1dugh/native-sound-mixer#41] [Q&A]
So I've been looking into this package for a while, wanting to add some features, these include session volume & mute event handling, although I got some questions/suggestions.
First off, I'm not really experienced with C++, memory management, or common good practices. 😅

Making that point clear, putting together the code for volume and mute event handling wasn't particularly hard with what's currently available in the code base, I only had to modify the JS source code and add a few overrides here and there, and it's working as intended, here is a screenshot of how it could be used:
![image](https://github.com/m1dugh/native-sound-mixer/assets/40004595/7ecf915f-ad7f-4221-9050-f79ab2d31bd1)

I have some questions regarding the EventPool object, currently, I'm using the same EventPool that is available on the mixer object, it comes with a few hindrances such as having to pass the audio session callback with the event pool reference to the session, and it all gets registered to one single event pool, I'm thinking of adding more events such as opening/closing new sessions to the event pool, so I can reload the available sessions in JS, register new events to them, and so on, but they will be new event pools with different methods and the such, I'm curious if I should, instead of binding everything to one single event pool, should I make an event pool for every Device/Session individually? Would it affect performance in some way? Any insight from you is greatly appreciated. 😁

**Comment by m1dugh** [comment]
The reason why the event pool is global is that we want the events to be bound to the id of the device rather than to the device object itself. That is because of the device object is destroyed in memory and then recreated later (which is the case when using a foreach for example), the idea is to still have the proper events bound to the device.

Nevertheless, the best option would be to have the event pool properly handling events separatly for devices and sessions which would still require to have the id of both the device and the audio session but would still satisfy the condition explained above.

With that explained, I guess it is exactly the code you have which I'd be really interested in seeing 😅.

**Comment by danielalejandromatheus** [comment]
You can view what I've done so far [here](https://github.com/m1dugh/native-sound-mixer/compare/master...danielalejandromatheus:native-sound-mixer:master), there's some noise in the javascript files because of my eslint configuration, but that would be the gist of it.

**Comment by m1dugh** [comment]
I'd like to keed the event pool with device descriptors instead of strings (even if it is the same at the end of the day). The way I would see this refactor of the event pools would be to have a tree in the event pool where the root would be the event pool, each first-level child would be an audio device with all its linked event callbacks and with additional children that would be all the audio sessions with registered events bound. 

This may not be very clear :sweat_smile:, I can create a branch on the repo with the implementation of my idea so you could see.

**Reply by danielalejandromatheus** [reply]
So far both audio and device events are on the same level of the EventPool, the only thing that I changed from the way you handle it is passing the ID of Devices/AudioSessions to your hashing function directly instead of the whole object to be able to reuse it, but other than that, each descriptor has their own method instead of only strings.
I think I can see what you mean by the event pool tree, it would be something like:
EventPool --> DeviceEvents --> SessionEvents including the remove/destroy events, although I'm not sure about the add/create events, as they would be on the root of the previous layer/branch based on how It works with the Windows API.

**Reply by m1dugh** [reply]
It would indeed make more sense to have the create/destroy events 1 layer above (root level for devices, and device level for sessions). Concerning the windows implementation, we can design the event pool for our needs and deal with windows API on the other side. Windows API should not be orienting the decision here.

Feel free to update your PR accordingly, or I will make the change myself as soon as I have time.

**Reply by danielalejandromatheus** [reply]
I'll update the PR this week with the new event pool structure, and then hopefully start working with the other events. 😁
